File: 11-container-with-most-water/11-container-with-most-water.rs
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut mh = 0;
        let mut it = height.into_iter().enumerate();
        let (mut left, mut right) = (it.next(), it.next_back());
        while let (Some((left_pos, left_height)), Some((right_pos, right_height))) = (left, right) {
            let base = (right_pos - left_pos) as i32;
            let min_height = match (left_height, right_height) {
                (height_left, height_right) if height_left < height_right => {
                    left = it.next();
                    height_left
                }
                (_, height_right) => {
                    right = it.next_back();
                    height_right
                }
            };
            mh = mh.max(base * min_height);
        }
        mh
    }
}

File: 11-container-with-most-water/11-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        ptr_a : int = 0
        ptr_b : int = len(height)-1
        
        result: int = 0 
        
        while ptr_b >= ptr_a:
            area = (ptr_b - ptr_a) * min(height[ptr_a], height[ptr_b])
            result = max(result,area)
            if height[ptr_a] > height[ptr_b]:
                ptr_b -= 1
            else:
                ptr_a += 1
            
        return result